infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/multiuser/oauth/OpenshiftTokenInitializationFilter.java

```diff
@@ -0,0 +1,152 @@
+import org.eclipse.che.multiuser.api.authentication.commons.token.RequestTokenExtractor;
+import org.eclipse.che.multiuser.api.permission.server.AuthorizedSubject;
+import org.eclipse.che.multiuser.api.permission.server.PermissionChecker;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This filter uses given token directly. It's used for native OpenShift user authentication.
+ * Requests without token or with invalid token are rejected.
+ *
+ * <p>{@link OpenshiftTokenInitializationFilter#UNAUTHORIZED_ENDPOINT_PATHS} is list of
+ * unauthenticated paths, that are allowed without token.
+ */
+@Singleton
+public class OpenshiftTokenInitializationFilter extends MultiUserEnvironmentInitializationFilter {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OpenshiftTokenInitializationFilter.class);
+
+  private static final List<String> UNAUTHORIZED_ENDPOINT_PATHS =
+      Collections.singletonList("/system/state");
+
+  private final PermissionChecker permissionChecker;
+  private final OpenShiftClientFactory clientFactory;
+
+  private final UserManager userManager;
+
+  @Inject
+  public OpenshiftTokenInitializationFilter(
+      SessionStore sessionStore,
+      RequestTokenExtractor tokenExtractor,
+      OpenShiftClientFactory clientFactory,
+      UserManager userManager,
+      PermissionChecker permissionChecker) {
+    super(sessionStore, tokenExtractor);
+    this.clientFactory = clientFactory;
+    this.userManager = userManager;
+    this.permissionChecker = permissionChecker;
+  }
+
+  @Override
+  protected String getUserId(String token) {
+    try {
+      return getCurrentUser(token).getMetadata().getUid();
+    } catch (KubernetesClientException e) {
+      if (e.getCode() == 401) {
+        LOG.error(
+            "Unauthorized when getting current user. Invalid OpenShift token, probably expired. Re-login? Re-request the token?");
+      }
+      throw new RuntimeException(e);
+    } catch (InfrastructureException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  protected Subject extractSubject(String token) {
+    try {
+      ObjectMeta userMeta = getCurrentUser(token).getMetadata();
+      User user =
+          userManager.getOrCreateUser(
+              userMeta.getUid(), openshiftUserEmail(userMeta), userMeta.getName());
+      return new AuthorizedSubject(
+          new SubjectImpl(user.getName(), user.getId(), token, false), permissionChecker);
+    } catch (InfrastructureException | ServerException | ConflictException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private io.fabric8.openshift.api.model.User getCurrentUser(String token)
+      throws InfrastructureException {
+    OpenShiftClient client = clientFactory.createAuthenticatedClient(token);
+    return client.currentUser();
+  }
```

Comment:
Can we get user-related information from token? Is there something missing? I thought it's jwt token like a keycloak token.

Reply:
afaik, the token format is not defined by oidc specification. Openshift does not use jwt token for user session, but only sha256 hash, see example here https://docs.openshift.com/container-platform/4.7/authentication/managing-oauth-access-tokens.html#oauth-view-details-tokens_managing-oauth-access-tokens

Reply:
How often we'll have to execute this part of the code? Once per HTTP(tomcat) session?

Reply:
Do you have in mind what to do on k8s? KubernetesClient doesn't have the `currentUser` method.

Reply:
this is a good question, but I don't have an answer for you. We need to take a look what OIDC guarantee for us and what options we have. We might get the info from the token, when we detect it's jwt or we might use the OIDC endpoint. Or some combination of both. Or maybe there is another option I'm not aware of.

Reply:
I think that's ok for the current state